app-commons/src/main/scala/org/bitcoins/commons/jsonmodels/dlc/OutcomeValueFunction.scala

```diff
@@ -0,0 +1,278 @@
+    case Some(firstMidpoint) =>
+      require(leftEndpoint.outcome < firstMidpoint.outcome,
+              s"Points must be ascending: $this")
+      require(midpoints.init.zip(midpoints.tail).forall {
+                case (m1, m2) => m1.outcome < m2.outcome
+              },
+              s"Points must be ascending: $this")
+      require(rightEndpoint.outcome > midpoints.last.outcome,
+              s"Points must be ascending: $this")
+    case None =>
+      require(leftEndpoint.outcome < rightEndpoint.outcome,
+              s"Points must be ascending: $this")
+  }
+
+  def apply(outcome: BigDecimal): Satoshis
+
+  def apply(outcome: BigDecimal, rounding: RoundingIntervals): Satoshis = {
+    rounding.round(outcome, apply(outcome))
+  }
+
+  /** Returns the largest Long less than or equal to bd (floor function) */
+  protected def bigDecimalSats(bd: BigDecimal): Satoshis = {
+    Satoshis(bd.setScale(0, RoundingMode.FLOOR).toLongExact)
+  }
+}
+
+object OutcomeValueFunctionComponent {
+
+  def apply(
+      points: Vector[OutcomeValuePoint]): OutcomeValueFunctionComponent = {
+    points match {
+      case Vector(left, right) =>
+        if (left.value == right.value) {
+          OutcomeValueConstant(left, right)
+        } else {
+          OutcomeValueLine(left, right)
+        }
+      case Vector(left, mid, right) => OutcomeValueQuadratic(left, mid, right)
+      case Vector(left, mid1, mid2, right) =>
+        OutcomeValueCubic(left, mid1, mid2, right)
+      case _ => OutcomeValuePolynomial(points)
+    }
+  }
+}
+
+case class OutcomeValueConstant(
+    leftEndpoint: OutcomeValuePoint,
+    rightEndpoint: OutcomeValuePoint)
+    extends OutcomeValueFunctionComponent {
+  require(leftEndpoint.value == rightEndpoint.value,
+          "Constant function must have same values on endpoints")
+
+  override def midpoints: Vector[OutcomeValuePoint] = Vector.empty
+
+  override def apply(outcome: BigDecimal): Satoshis = leftEndpoint.value
+}
+
+/** A Line between left and right endpoints defining a piece of a larger payout curve */
+case class OutcomeValueLine(
+    leftEndpoint: OutcomeValuePoint,
+    rightEndpoint: OutcomeValuePoint)
+    extends OutcomeValueFunctionComponent {
+  override def midpoints: Vector[OutcomeValuePoint] = Vector.empty
+
+  lazy val slope: BigDecimal = {
+    (rightEndpoint.value.toLong - leftEndpoint.value.toLong) / (rightEndpoint.outcome - leftEndpoint.outcome)
+  }
+
+  override def apply(outcome: BigDecimal): Satoshis = {
+    val value =
+      (outcome - leftEndpoint.outcome) * slope + leftEndpoint.value.toLong
+
+    bigDecimalSats(value)
+  }
+}
+
+/** A quadratic between left and right endpoints defining a piece of a larger payout curve */
+case class OutcomeValueQuadratic(
+    leftEndpoint: OutcomeValuePoint,
+    midpoint: OutcomeValuePoint,
+    rightEndpoint: OutcomeValuePoint)
+    extends OutcomeValueFunctionComponent {
+  override def midpoints: Vector[OutcomeValuePoint] = Vector(midpoint)
```

Comment:
nit: Probably worth making a `val` otherwise a `Vector()` gets allocated everytime this is called.

Reply:
This applies elsewhere in this file too.